BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,437 @@
+            cs.y = elem->FloatAttribute("csy");
+            //this option available only for spot light
+            SpotLight::Ptr spot = std::dynamic_pointer_cast<SpotLight>(new_light);
+            spot->SetConeShape(cs);
+        }
+        else if (type == "ibl")
+        {
+            new_light = ImageBasedLight::Create();
+            std::string tex_name = elem->Attribute("tex");
+            float mul = elem->FloatAttribute("mul");
+            
+            //this option available only for ibl
+            ImageBasedLight::Ptr ibl = std::dynamic_pointer_cast<ImageBasedLight>(new_light);
+            auto image_io = ImageIo::CreateImageIo();
+            Texture::Ptr tex = image_io->LoadImage(tex_name.c_str());
+            ibl->SetTexture(tex);
+            ibl->SetMultiplier(mul);
+        }
+        else
+        {
+            throw std::runtime_error("Render::LoadLightXml(...): Invalid light type " + type);
+        }
+
+        RadeonRays::float3 p;
+        RadeonRays::float3 d;
+        RadeonRays::float3 r;
+
+        p.x = elem->FloatAttribute("posx");
+        p.y = elem->FloatAttribute("posy");
+        p.z = elem->FloatAttribute("posz");
+
+        d.x = elem->FloatAttribute("dirx");
+        d.y = elem->FloatAttribute("diry");
+        d.z = elem->FloatAttribute("dirz");
+
+        r.x = elem->FloatAttribute("radx");
+        r.y = elem->FloatAttribute("rady");
+        r.z = elem->FloatAttribute("radz");
+
+        new_light->SetPosition(p);
+        new_light->SetDirection(d);
+        new_light->SetEmittedRadiance(r);
+        m_scene->AttachLight(new_light);
+        elem = elem->NextSiblingElement("light");
+    }
+}
+
+void Render::UpdateCameraSettings(const CameraInfo& cam_state)
+{
+    if (cam_state.aperture != m_camera->GetAperture())
+    {
+        m_camera->SetAperture(cam_state.aperture);
+    }
+
+    if (cam_state.focal_length != m_camera->GetFocalLength())
+    {
+        m_camera->SetFocalLength(cam_state.focal_length);
+    }
+
+    if (cam_state.focus_distance != m_camera->GetFocusDistance())
+    {
+        m_camera->SetFocusDistance(cam_state.focus_distance);
+    }
+
+    auto cur_pos = m_camera->GetPosition();
+    auto at = m_camera->GetForwardVector();
+    auto up = m_camera->GetUpVector();
+
+    if (cur_pos != cam_state.pos ||
+        at != cam_state.at ||
+        up != cam_state.up)
+    {
+        m_camera->LookAt(cam_state.pos, cam_state.at, cam_state.up);
+    }
+}
+
+void Render::SaveOutput(Renderer::OutputType type,
+                        const std::string& path,
+                        const std::string& file_name,
+                        const std::string& extension,
+                        int bpp)
+{
+    OIIO_NAMESPACE_USING;
+
+    std::vector<RadeonRays::float3> output_data;
+    std::vector<RadeonRays::float3> image_data;
+    auto output = m_renderer->GetOutput(type);
+    auto width = output->width();
+    auto height = output->height();
```

Comment:
lets move width and height initialization below (to place of first usage)

Reply:
done